File: libs/document-processor-next/src/engine/processor-execution-context.ts
import type { ContractBundle } from './contract-bundle.js';
import type { JsonPatch } from '../model/shared/json-patch.js';
import { DocumentProcessingRuntime } from '../runtime/document-processing-runtime.js';
import { ProcessorFatalError } from './processor-fatal-error.js';
import { Blue, BlueNode } from '@blue-labs/language';
import { ProcessorErrors } from '../types/errors.js';
import { ProcessorEngine } from './processor-engine.js';

export interface ExecutionAdapter {
  runtime(): DocumentProcessingRuntime;
  isScopeInactive(scopePath: string): boolean;
  handlePatch(
    scopePath: string,
    bundle: ContractBundle,
    patch: JsonPatch,
    allowReservedMutation: boolean,
  ): void;
  resolvePointer(scopePath: string, relativePointer: string): string;
  enterGracefulTermination(
    scopePath: string,
    bundle: ContractBundle,
    reason: string | null,
  ): void;
  enterFatalTermination(
    scopePath: string,
    bundle: ContractBundle,
    reason: string | null,
  ): void;
}

export class ProcessorExecutionContext {
  constructor(
    private readonly execution: ExecutionAdapter,
    private readonly bundle: ContractBundle,
    private readonly scopePathValue: string,
    private readonly eventNode: BlueNode,
    private readonly allowTerminatedWork: boolean,
    private readonly allowReservedMutation: boolean,
  ) {}

  get scopePath(): string {
    return this.scopePathValue;
  }

  get blue(): Blue {
    return this.execution.runtime().blue();
  }

  event(): BlueNode {
    return this.eventNode;
  }

  applyPatch(patch: JsonPatch): void {
    if (
      !this.allowTerminatedWork &&
      this.execution.isScopeInactive(this.scopePathValue)
    ) {
      return;
    }
    this.execution.handlePatch(
      this.scopePathValue,
      this.bundle,
      patch,
      this.allowReservedMutation,
    );
  }

  emitEvent(emission: BlueNode): void {
    if (
      !this.allowTerminatedWork &&
      this.execution.isScopeInactive(this.scopePathValue)
    ) {
      return;
    }
    const runtime = this.execution.runtime();
    const scopeContext = runtime.scope(this.scopePathValue);
    runtime.chargeEmitEvent(emission);
    const queued = emission.clone();
    scopeContext.enqueueTriggered(queued);
    scopeContext.recordBridgeable(queued.clone());
    if (this.scopePathValue === '/') {
      runtime.recordRootEmission(queued.clone());
    }
  }

  consumeGas(units: number): void {
    if (
      !this.allowTerminatedWork &&
      this.execution.isScopeInactive(this.scopePathValue)
    ) {
      return;
    }
    this.execution.runtime().addGas(units);
  }

  throwFatal(reason: string): never {
    throw new ProcessorFatalError(reason, ProcessorErrors.runtimeFatal(reason));
  }

  resolvePointer(relativePointer: string): string {
    return this.execution.resolvePointer(this.scopePathValue, relativePointer);
  }

  documentAt(absolutePointer: string): BlueNode | null {
    if (!absolutePointer) {
      return null;
    }
    try {
      const node = this.documentNodeAt(absolutePointer);
      return node ? node.clone() : null;
    } catch {
      return null;
    }
  }

  documentContains(absolutePointer: string): boolean {
    if (!absolutePointer) {
      return false;
    }
    try {
      return (
        ProcessorEngine.nodeAt(
          this.execution.runtime().document(),
          absolutePointer,
        ) != null
      );
    } catch {
      return false;
    }
  }

  private documentNodeAt(absolutePointer: string): BlueNode | null {
    const node = ProcessorEngine.nodeAt(
      this.execution.runtime().document(),
      absolutePointer,
    );
    return node instanceof BlueNode ? node : null;
  }

  terminateGracefully(reason: string | null): void {
    this.execution.enterGracefulTermination(
      this.scopePathValue,
      this.bundle,
      reason ?? null,
    );
  }

  terminateFatally(reason: string | null): void {
    this.execution.enterFatalTermination(
      this.scopePathValue,
      this.bundle,
      reason ?? null,
    );
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/ProcessorExecutionContext.java
package blue.language.processor;

import blue.language.NodeProvider;
import blue.language.model.Node;
import blue.language.processor.util.PointerUtils;
import blue.language.processor.model.JsonPatch;

import java.math.BigInteger;
import java.util.Objects;

/**
 * Lightweight wrapper passed to contract processors while executing.
 */
public final class ProcessorExecutionContext {

    private final ProcessorEngine.Execution execution;
    private final ContractBundle bundle;
    private final String scopePath;
    private final Node event;
    private final boolean allowTerminatedWork;
    private final boolean allowReservedMutation;

    ProcessorExecutionContext(ProcessorEngine.Execution execution,
                              ContractBundle bundle,
                              String scopePath,
                              Node event,
                              boolean allowTerminatedWork,
                              boolean allowReservedMutation) {
        this.execution = Objects.requireNonNull(execution, "execution");
        this.bundle = Objects.requireNonNull(bundle, "bundle");
        this.scopePath = Objects.requireNonNull(scopePath, "scopePath");
        this.event = Objects.requireNonNull(event, "event");
        this.allowTerminatedWork = allowTerminatedWork;
        this.allowReservedMutation = allowReservedMutation;
    }

    public Node event() {
        return event;
    }

    public void applyPatch(JsonPatch patch) {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        execution.handlePatch(scopePath, bundle, patch, allowReservedMutation, allowTerminatedWork);
    }

    public void emitEvent(Node emission) {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        Objects.requireNonNull(emission, "emission");
        DocumentProcessingRuntime runtime = runtime();
        ScopeRuntimeContext scopeContext = runtime.scope(scopePath);
        runtime.chargeEmitEvent(emission);
        Node queued = emission.clone();
        scopeContext.enqueueTriggered(queued);
        scopeContext.recordBridgeable(queued.clone());
        if ("/".equals(scopeContext.scopePath())) {
            runtime.recordRootEmission(queued.clone());
        }
    }

    public void consumeGas(long units) {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        runtime().addGas(units);
    }

    public void chargeTriggerEventBase() {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        runtime().chargeTriggerEventBase();
    }

    public void chargeUpdateDocumentBase(int changesetLength) {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        runtime().chargeUpdateDocumentBase(changesetLength);
    }

    public void chargeDocumentSnapshot(String absolutePointer, Node snapshot) {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        runtime().chargeDocumentSnapshot(absolutePointer, snapshot);
    }

    public void chargeWasmGas(BigInteger wasmFuel) {
        if (!allowTerminatedWork && execution.isScopeInactive(scopePath)) {
            return;
        }
        runtime().chargeWasmGas(wasmFuel);
    }

    public void throwFatal(String reason) {
        throw new ProcessorFatalException(reason);
    }

    public String resolvePointer(String pointer) {
        return PointerUtils.resolvePointer(scopePath, pointer);
    }

    public Node documentAt(String absolutePointer) {
        if (absolutePointer == null || absolutePointer.trim().isEmpty()) {
            return null;
        }
        try {
            String normalizedPointer = PointerUtils.normalizePointer(absolutePointer);
            Node node = ProcessorEngine.nodeAt(runtime().document(), normalizedPointer);
            return node != null ? node.clone() : null;
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    public boolean documentContains(String absolutePointer) {
        if (absolutePointer == null || absolutePointer.trim().isEmpty()) {
            return false;
        }
        try {
            String normalizedPointer = PointerUtils.normalizePointer(absolutePointer);
            Node node = ProcessorEngine.nodeAt(runtime().document(), normalizedPointer);
            return node != null;
        } catch (RuntimeException ignored) {
            return false;
        }
    }

    public void terminateGracefully(String reason) {
        execution.enterGracefulTermination(scopePath, bundle, reason);
    }

    public void terminateFatally(String reason) {
        execution.enterFatalTermination(scopePath, bundle, reason);
    }

    public NodeProvider nodeProvider() {
        DocumentProcessor owner = execution.owner();
        if (owner == null || owner.registry() == null) {
            return null;
        }
        return owner.registry().nodeProvider();
    }

    private DocumentProcessingRuntime runtime() {
        return execution.runtime();
    }
}
